server/monitor/src/main/java/org/apache/accumulo/monitor/util/AccumuloMonitorAppender.java

```diff
@@ -0,0 +1,153 @@
+    trackerScheduled = new AtomicBoolean(false);
+    executorService = Executors.newSingleThreadScheduledExecutor(runnable -> {
+      Thread t = new Thread(runnable, AccumuloMonitorAppender.class.getSimpleName() + " Location Tracker");
+      t.setDaemon(true);
+      return t;
+    });
+  }
+
+  @Override
+  public void activateOptions() {
+    // only schedule it once (in case options get activated more than once); not sure if this is possible
+    if (trackerScheduled.compareAndSet(false, true)) {
+      // wait 5 seconds, then run every 5 seconds
+      executorService.scheduleAtFixedRate(new MonitorTracker(), 5, 5, TimeUnit.SECONDS);
+    }
+    super.activateOptions();
+  }
+
+  @Override
+  public void close() {
+    if (!executorService.isShutdown()) {
+      executorService.shutdownNow();
+    }
+    super.close();
+  }
+
+  private class MonitorTracker implements Runnable {
+
+    private String path;
+    private ZooCache zooCache;
+
+    private long lastModifiedTransactionId;
+    private SocketAppender lastSocketAppender;
+
+    public MonitorTracker() {
+
+      // path and zooCache are lazily set the first time this tracker is run
+      // this allows the tracker to be constructed and scheduled during log4j initialization without
+      // triggering any actual logs from the Accumulo or ZooKeeper code
+      this.path = null;
+      this.zooCache = null;
+
+      this.lastModifiedTransactionId = 0;
+      this.lastSocketAppender = null;
+    }
+
+    @Override
+    public void run() {
+      try {
+        // lazily set up path and zooCache (see comment in constructor)
+        if (this.zooCache == null) {
+          Instance instance = HdfsZooInstance.getInstance();
+          this.path = ZooUtil.getRoot(instance) + Constants.ZMONITOR_LOG4J_ADDR;
+          this.zooCache = new ZooCacheFactory().getZooCache(instance.getZooKeepers(), instance.getZooKeepersSessionTimeOut());
+        }
+
+        // get the current location from the cache and update if necessary
+        Stat stat = new Stat();
+        byte[] loc = zooCache.get(path, stat);
+        long modifiedTransactionId = stat.getMzxid();
+
+        // modifiedTransactionId will be 0 if no current location
+        // lastModifiedTransactionId will be 0 if we've never seen a location
+        if (modifiedTransactionId != lastModifiedTransactionId) {
+          // replace old socket on every change, even if new location is the same as old location
+          // if modifiedTransactionId changed, then the monitor restarted and the old socket is dead now
+          switchAppender(loc, modifiedTransactionId);
+        }
+      } catch (Exception e) {
+        // dump any non-fatal problems to the console, but let it run again
+        e.printStackTrace();
+      }
+    }
+
+    private void switchAppender(byte[] newLocation, long newModifiedTransactionId) {
+      // remove and close the last one, if it was non-null
+      if (lastSocketAppender != null) {
```

Comment:
It won't be called multiple times with null. If `newLocation` is null, it means the ZK node didn't exist, and the modified transaction ID will be 0. So the code outside guards against, because the last modified transaction id is updated to 0 the first time `newLocation` is null.

However, even if it were called multiple times, both the remove of the last appender and the close are both NOOPs if it has already been removed and closed.